api/src/generate-service-worker.js

```diff
@@ -0,0 +1,137 @@
+const swPrecache = require('sw-precache');
+const path = require('path');
+const fs = require('fs');
+const crypto = require('crypto');
+
+const environment = require('./environment');
+const db = require('./db');
+const logger = require('./logger');
+
+const SWMETA_DOC_ID = 'service-worker-meta';
+const apiSrcDirectoryPath = __dirname;
+
+const staticDirectoryPath = environment.getExtractedResourcesPath();
+const scriptOutputPath = path.join(staticDirectoryPath, 'js', 'service-worker.js');
+
+/**
+ * Generates a sha1 of the existent service worker file.
+ * If the file does not exist, it returns undefined.
+ * If any error occurs when reading, it returns undefined.
+ * If any error occurs when generating the hash, it returns undefined.
+ * @return {Promise<string|undefined>}
+ */
+const getServiceWorkerHash = () => {
+  return new Promise((resolve) => {
+    try {
+      if (!fs.existsSync(scriptOutputPath)) {
+        return resolve();
+      }
+
+      const hash = crypto.createHash('sha1');
+      const stream = fs.createReadStream(scriptOutputPath);
+      stream.setEncoding('utf8');
+      stream.on('data', (data) => hash.update(data));
+      stream.on('end', () => resolve(hash.digest('hex')));
+      stream.on('error', (err) => {
+        logger.error('Error while reading service worker: %o', err);
+        resolve();
+      });
+    } catch (err) {
```

Comment:
This is a part I'm not certain about. 

I thought that the service worker is a key piece of our setup and we should terminate API if generating it fails - like we terminate when we fail to extract resources or fail load settings. 
I wanted to make sure that non-critical errors (like crypto errors of any kind) don't end up terminating API, but are still logged and visible. 
My conclusion was that we should only terminate (have this module throw) when actually writing the service worker file to disk fails, and catch everything else. 
I'm reluctant about terminating the process when we update the sw meta file. It's also critical for users to get the new code when we upgrade, but if Couch is under load and is timing out requests ... we probably shouldn't terminate? 

I'd appreciate your opinion.

Reply:
I agree we shouldn't terminate API. If I understand the proposal here then this function will resolve `undefined` which will then get written to the SW doc. This will prompt every client to try and update their SW which could brick the phones if there's something wrong with the extracted files, right? I'm wondering if we want to change the logic so if the new hash is `undefined` it's a no-op, and we rely on it working next time.

Reply:
Done. No longer updating the meta doc if the new hash fails.